mujoco_py/tests/test_substep.py

```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python
+import unittest
+from mujoco_py import load_model_from_xml, MjSim
+
+XML = '''
+<mujoco>
+    <size nuserdata="{nuserdata}"/>
+    <worldbody>
+        <body pos="0 0 0">
+            <joint name="j1" type="hinge" axis="1 0 0"/>
+            <geom type="sphere" size=".5"/>
+            <body pos="0 0 1">
+                <joint name="j2" type="hinge" axis="1 0 0"/>
+                <geom type="sphere" size=".5"/>
+            </body>
+        </body>
+    </worldbody>
+    <actuator>
+        <position joint="j1" kp="100"/>
+        <position joint="j2" kp="100"/>
+    </actuator>
+</mujoco>
+'''
+
+HELLO_FN = '''
+    #include <stdio.h>
+    void generic(const mjModel* m, mjData* d) {
+        printf("hello");
+    }
+'''
+
+INCREMENT_FN = '''
+    void generic(const mjModel* m, mjData* d) {
+        d->userdata[0] += 1;
+    }
+'''
+
+
+class TestSubstep(unittest.TestCase):
+    def test_hello(self):
+        sim = MjSim(load_model_from_xml(XML.format(nuserdata=0)),
+                    substep_udd_fn=HELLO_FN)
+        sim.step()  # should print 'hello'
+
+    def test_increment(self):
+        sim = MjSim(load_model_from_xml(XML.format(nuserdata=1)),
+                    substep_udd_fn=INCREMENT_FN)
+        self.assertEqual(sim.data.userdata[0], 0)
+        sim.step()  # should increment userdata[0]
+        self.assertEqual(sim.data.userdata[0], 1)
+        # Test again with different nsubsteps, reuse model
+        sim = MjSim(sim.model, nsubsteps=7, substep_udd_fn=INCREMENT_FN)
+        self.assertEqual(sim.data.userdata[0], 0)
+        sim.step()  # should increment userdata[0] 7 times
+        self.assertEqual(sim.data.userdata[0], 7)
+
+    def test_sum_ctrl(self):
```

Comment:
This test (`test_sum_ctrl`) is probably the best standalone example of what it looks like to use this.